File: 08_04_Power_Set.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        auto n = nums.size();
        auto pow = 1 << n;

        vector<vector<int>>subSets(1, vector<int>());
        for(auto i = 1; i < pow; i++) {
            vector<int> currentSet;
            for(auto j = 0; j < nums.size(); j++) {
                if (((i >> j) & 1) > 0)
                    currentSet.push_back(nums[j]);
            }
            subSets.push_back(currentSet);
        }        
        return subSets;
    }
};

File: 08_04_Power_Set.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        
        List<List<Integer>> reVal = new ArrayList<>();
        
        for (int i = 0;i < Math.pow(2, nums.length);i++) {
            int bits = i;
            List<Integer> sublist = new ArrayList<>();
            int atBinaryDigit = 0;
            while (bits > 0) {
                if ((1 & bits) == 1) {
                    sublist.add(nums[atBinaryDigit]);
                }
                atBinaryDigit++;
                bits >>>= 1;
            }
            reVal.add(sublist);
            
        }
        
        return reVal;
    }
}
